### PR TITLE
test: ecrecover evm compatibility

### LINKED ISSUE #814: Ecrecover EVM compatibility
### Problem

The behavior of the ecrecover function with various keys is sometimes not clear. It would be good to provide developers with some examples.

Accounts with an ECDSA key pair and an EVM alias derived from them should be able to use the `ecrecover` function correctly, provided their key pair has not been replaced.

In the Hedera Network:
- `msg.sender` always returns the EVM alias associated with the user account, even if the key pairs associated with that account are replaced. This means `msg.sender` will remain consistent for every transaction signed by that account.
- The `ecrecover` function, however, returns the address associated with the actual private key used to sign the message passed to it. Therefore, after the first key pair update, `ecrecover` will no longer work correctly for ECDSA accounts. Despite this, `msg.sender` will remain unchanged and consistent.

Please note that `ecrecover` will never work for ED25519 accounts, regardless of the circumstances.

### Solution

Provide a batch of tests for ecrecover function with the usage of different keys.

### Alternatives

_No response_
```[tasklist]
### Tasks
- [ ] https://github.com/hashgraph/hedera-smart-contracts/pull/815
```

### PR DESCRIPTION
**Pull request by marcin-piela-ariane**
**Description**:

This PR adds tests and documents the behavior of the ecrecover function with various keys, providing developers with clear examples of when ecrecover works correctly.

**Related issue(s)**:

Closes #814 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Comment by quiet-node** [comment]
Great work but please include artifact files.

**Review by quiet-node — CHANGES_REQUESTED** [review]
Great work but please include artifact files.

**Comment by marcin-piela-ariane** [comment]
> Great work but please include artifact files.

@quiet-node Sorry for that. Added ABI & adjusted to the new directory structure.

**Review by quiet-node — CHANGES_REQUESTED** [review]
Looks like you're targeting the merge against `hashgraph:main` branch, instead of `main`. Is this on purpose?

**Comment by marcin-piela-ariane** [comment]
> Looks like you're targeting the merge against `hashgraph:main` branch, instead of `main`. Is this on purpose?

@quiet-node Yes, I'd like to contribute to the main project from my forked repository (branch `test-ecrecover-evm-compatibility`).

**Comment by quiet-node** [comment]
> > Looks like you're targeting the merge against `hashgraph:main` branch, instead of `main`. Is this on purpose?
> 
> @quiet-node Yes, I'd like to contribute to the main project from my forked repository (branch `test-ecrecover-evm-compatibility`).

Oh ah yeah sorry I completely forgot they were technically the same branch. 😓

**Review by quiet-node — COMMENTED** [review]
I tried running the test and none of them were working...

```bash
╰─ npx hardhat test                                                                                                               ─╯


  EcrecoverCheck
    Deployment
      1) Should be deployed correctly
    Verification
      2) Ecrecover should work correctly for account with ECDSA key and EVM alias derived from ECDSA key.
      3) Ecrecover should fail for account with ECDSA key replaced by new ECDSA private key. EVMAlias (msg.sender) will remain the same but signer extracted with ecrecover will be derived from new key pair.
      4) Ecrecover should fail for account with ECDSA key replaced by new ED25519 private key. EVMAlias (msg.sender) will remain the same but signer extracted with ecrecover will be some random value, because ecrecover will not work for ED25519 keys.
      5) Ecrecover should be broken for account with ECDSA key and default EVM alias. EVM alias is not connected in any way to the ECDSA key, so ecrecover result will not return it.
      6) Ecrecover should be broken for ED25519 keys. No matter what they will be replaced with.


  0 passing (2s)
  6 failing

  1) EcrecoverCheck
       Deployment
         Should be deployed correctly:
     AssertionError: expected null not to be null
      at Context.<anonymous> (test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:61:29)
      at processImmediate (node:internal/timers:476:21)

  2) EcrecoverCheck
       Verification
         Ecrecover should work correctly for account with ECDSA key and EVM alias derived from ECDSA key.:
     StatusError: transaction 0.0.2@1719347061.872247611 failed precheck with status INVALID_SIGNATURE
      at new PrecheckStatusError (node_modules/@hashgraph/sdk/lib/PrecheckStatusError.cjs:50:5)
      at AccountCreateTransaction._mapStatusError (node_modules/@hashgraph/sdk/lib/transaction/Transaction.cjs:1320:12)
      at AccountCreateTransaction.execute (node_modules/@hashgraph/sdk/lib/Executable.cjs:667:22)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
      at Function.createAccount (test/precompile/native/evm-compatibility-ecrecover/utils.js:69:22)
      at testEcrecover (test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:34:19)
      at Context.<anonymous> (test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:67:14)

  3) EcrecoverCheck
       Verification
         Ecrecover should fail for account with ECDSA key replaced by new ECDSA private key. EVMAlias (msg.sender) will remain the same but signer extracted with ecrecover will be derived from new key pair.:
     StatusError: transaction 0.0.2@1719347062.118677236 failed precheck with status INVALID_SIGNATURE
      at new PrecheckStatusError (node_modules/@hashgraph/sdk/lib/PrecheckStatusError.cjs:50:5)
      at AccountCreateTransaction._mapStatusError (node_modules/@hashgraph/sdk/lib/transaction/Transaction.cjs:1320:12)
      at AccountCreateTransaction.execute (node_modules/@hashgraph/sdk/lib/Executable.cjs:667:22)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
      at Function.createAccount (test/precompile/native/evm-compatibility-ecrecover/utils.js:69:22)
      at testEcrecover (test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:34:19)
      at Context.<anonymous> (test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:71:14)

  4) EcrecoverCheck
       Verification
         Ecrecover should fail for account with ECDSA key replaced by new ED25519 private key. EVMAlias (msg.sender) will remain the same but signer extracted with ecrecover will be some random value, because ecrecover will not work for ED25519 keys.:
     StatusError: transaction 0.0.2@1719347063.719160638 failed precheck with status INVALID_SIGNATURE
      at new PrecheckStatusError (node_modules/@hashgraph/sdk/lib/PrecheckStatusError.cjs:50:5)
      at AccountCreateTransaction._mapStatusError (node_modules/@hashgraph/sdk/lib/transaction/Transaction.cjs:1320:12)
      at AccountCreateTransaction.execute (node_modules/@hashgraph/sdk/lib/Executable.cjs:667:22)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
      at Function.createAccount (test/precompile/native/evm-compatibility-ecrecover/utils.js:69:22)
      at testEcrecover (test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:34:19)
      at Context.<anonymous> (test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:75:14)

  5) EcrecoverCheck
       Verification
         Ecrecover should be broken for account with ECDSA key and default EVM alias. EVM alias is not connected in any way to the ECDSA key, so ecrecover result will not return it.:
     StatusError: transaction 0.0.2@1719347061.751742220 failed precheck with status INVALID_SIGNATURE
      at new PrecheckStatusError (node_modules/@hashgraph/sdk/lib/PrecheckStatusError.cjs:50:5)
      at AccountCreateTransaction._mapStatusError (node_modules/@hashgraph/sdk/lib/transaction/Transaction.cjs:1320:12)
      at AccountCreateTransaction.execute (node_modules/@hashgraph/sdk/lib/Executable.cjs:667:22)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
      at Function.createAccount (test/precompile/native/evm-compatibility-ecrecover/utils.js:69:22)
      at testEcrecover (test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:34:19)
      at Context.<anonymous> (test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:79:14)

  6) EcrecoverCheck
       Verification
         Ecrecover should be broken for ED25519 keys. No matter what they will be replaced with.:
     StatusError: transaction 0.0.2@1719347061.558526821 failed precheck with status INVALID_SIGNATURE
      at new PrecheckStatusError (node_modules/@hashgraph/sdk/lib/PrecheckStatusError.cjs:50:5)
      at AccountCreateTransaction._mapStatusError (node_modules/@hashgraph/sdk/lib/transaction/Transaction.cjs:1320:12)
      at AccountCreateTransaction.execute (node_modules/@hashgraph/sdk/lib/Executable.cjs:667:22)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
      at Function.createAccount (test/precompile/native/evm-compatibility-ecrecover/utils.js:69:22)
      at testEcrecover (test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:34:19)
      at Context.<anonymous> (test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:83:14)
 ```

**Comment by marcin-piela-ariane** [comment]
@quiet-node 

> I tried running the test and none of them were working...

Did you set `OPERATOR_ID_A` & `PRIVATE_KEYS ` as ECDSA keys?

> Important: While Hedera supports both ECDSA and ED25519 accounts, please use ECDSA since Ethereum only supports ECDSA.
(source: TEST_SETUP.md)

**Comment by quiet-node** [comment]
> @quiet-node
> 
> > I tried running the test and none of them were working...
> 
> Did you set `OPERATOR_ID_A` & `PRIVATE_KEYS ` as ECDSA keys?
> 
> > Important: While Hedera supports both ECDSA and ED25519 accounts, please use ECDSA since Ethereum only supports ECDSA.
> > (source: TEST_SETUP.md)

Ah yeah I'm aware of those options and I have been using EDCSA. I just switched to different environment, testnet, and it works fine. However, none of them were passing on localnet.

**Comment by marcin-piela-ariane** [comment]
> > @quiet-node
> > > I tried running the test and none of them were working...
> > 
> > 
> > Did you set `OPERATOR_ID_A` & `PRIVATE_KEYS ` as ECDSA keys?
> > > Important: While Hedera supports both ECDSA and ED25519 accounts, please use ECDSA since Ethereum only supports ECDSA.
> > > (source: TEST_SETUP.md)
> 
> Ah yeah I'm aware of those options and I have been using EDCSA. I just switched to different environment, testnet, and it works fine. However, none of them were passing on localnet.

@quiet-node OK. Thanks. I have to admit I didn't test it on local node. Only on testnet. Will have to dig into it.

**Comment by marcin-piela-ariane** [comment]
@quiet-node Fixed for local node testing. Wrong mirror node URL was the problem. Also added a retry mechanism for mirror node querying. Should be good to go now.

**Review by Nana-EC — COMMENTED** [review]
Some suggestions

File: contracts/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.sol [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.5.0 <0.9.0;
+
+function itoa(uint value) pure returns (string memory) {
```
[the remark this answers is not in the document]
Comment (arianejasuwienas) [reply contracts/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.sol]: @Nana-EC 
I’ve renamed the function to uintToString and added comments for better clarity. Let me know if there’s anything else that needs attention.

This function:
 - calculates the length of the string representation,
 - creates a bytes array with the calculated length,
 - fills the bytes array with the ASCII values of the digits (and returns it as a result eventually).

Should I include it in the comments in the function body as well, or is it clear by looking at the code?

File: test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  describe('Deployment', function () {
+    it('Should be deployed correctly', async function () {
+      expect(address).to.not.null;
+    });
+  });
+
+  describe('Verification', function () {
+    it('Ecrecover should work correctly for account with ECDSA key and EVM alias derived from ECDSA key.', async function () {
+      expect(await testEcrecover(address, 'ECDSA', true, '')).to.true;
```
Comment (acuarica) [inline test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:67]: Could you convert these `to.true`/`to.false` to address comparison directly? This is to provide more contextual information when the test fails.

Essentially, perform the test of `utils.verifyEcrecover` directly in the `expect`.

**Review by acuarica — CHANGES_REQUESTED** [review]
lg, some nits

File: test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    });
+
+    it('Ecrecover should fail for account with ECDSA key replaced by new ECDSA private key. EVMAlias (msg.sender) will remain the same but signer extracted with ecrecover will be derived from new key pair.', async function () {
```
[the remark this answers is not in the document]
Comment (arianejasuwienas) [reply test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js:78]: @Nana-EC I’ve implemented this scenario. After reverting to the initial private key, ecrecover works correctly again.

File: test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js (hunk 2 above)
Comment (arianejasuwienas) [reply test/precompile/native/evm-compatibility-ecrecover/EcrecoverCheck.js]: @acuarica I have implemented the requested changes. String comparison is now utilized in the test cases.

**Review by acuarica — APPROVED** [review]
lgtm

**Comment by se7enarianelabs** [comment]
@quiet-node May You resolve Your change request? 
https://github.com/hashgraph/hedera-smart-contracts/pull/815#pullrequestreview-2136193297

**Review by quiet-node — APPROVED** [review]
LGTM. Tested locally and it work as expected! Thanks!

**Pull request merged**